execute/report/report_test.go

```diff
@@ -199,24 +408,288 @@ func Test_buildSingleChainReport_Errors(t *testing.T) {
+						cciptypes.Bytes32{}, // generate a correct root.
+						[]cciptypes.SeqNum{100, 102, 104, 106, 108}),
+				},
+			},
+			expectedExecReports:   1,
+			expectedCommitReports: 1,
+			expectedExecThings:    []int{4},
+			lastReportExecuted:    []cciptypes.SeqNum{100, 101, 102, 103, 104, 105, 106, 107, 108},
+		},
+		{
+			name: "broken report",
+			args: args{
+				maxReportSize: 10000,
+				reports: []plugintypes.ExecutePluginCommitDataWithMessages{
+					breakCommitReport(makeTestCommitReport(hasher, 10, 1, 101, 1000, 10101010102,
+						cciptypes.Bytes32{}, // generate a correct root.
+						nil)),
+				},
+			},
+			wantErr: "unable to add a single chain report",
+		},
+		{
+			name: "invalid merkle root",
+			args: args{
+				reports: []plugintypes.ExecutePluginCommitDataWithMessages{
+					makeTestCommitReport(hasher, 10, 1, 100, 999, 10101010101,
+						mustMakeBytes(""), // random root
+						nil),
+				},
+			},
+			wantErr: "merkle root mismatch: expected 0x00000000000000000",
+		},
+		{
+			name: "skip over one large messages",
+			args: args{
+				maxReportSize: 10000,
+				reports: []plugintypes.ExecutePluginCommitDataWithMessages{
+					setMessageData(5, 20000,
+						makeTestCommitReport(hasher, 10, 1, 100, 999, 10101010101,
+							cciptypes.Bytes32{}, // generate a correct root.
+							nil)),
+				},
+			},
+			expectedExecReports:   1,
+			expectedCommitReports: 0,
+			expectedExecThings:    []int{9},
+			lastReportExecuted:    []cciptypes.SeqNum{100, 101, 102, 103, 104, 106, 107, 108, 109},
+		},
+		{
+			name: "skip over two large messages",
+			args: args{
+				maxReportSize: 10000,
+				reports: []plugintypes.ExecutePluginCommitDataWithMessages{
+					setMessageData(8, 20000,
+						setMessageData(5, 20000,
+							makeTestCommitReport(hasher, 10, 1, 100, 999, 10101010101,
+								cciptypes.Bytes32{}, // generate a correct root.
+								nil))),
+				},
+			},
+			expectedExecReports:   1,
+			expectedCommitReports: 0,
+			expectedExecThings:    []int{8},
+			lastReportExecuted:    []cciptypes.SeqNum{100, 101, 102, 103, 104, 106, 107, 109},
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
 			ctx := context.Background()
-			execReport, size, err := buildSingleChainReport(
-				ctx, lggr, resolvedHasher, resolvedTokenDataReader, resolvedCodec, tt.args.report, 0)
+			// look for error in Add or Build
+			foundError := false
+
+			builder := NewBuilder(ctx, lggr, hasher, tokenDataReader, codec, tt.args.maxReportSize, 0)
+			var updatedMessages []plugintypes.ExecutePluginCommitDataWithMessages
+			for _, report := range tt.args.reports {
+				updatedMessage, err := builder.Add(report)
+				if err != nil && tt.wantErr != "" {
+					if strings.Contains(err.Error(), tt.wantErr) {
+						foundError = true
+						break
+					}
+				}
+				require.NoError(t, err)
+				updatedMessages = append(updatedMessages, updatedMessage)
+			}
```

Comment:
Except for this part that swaps `selectReports` with the new builder, this test is almost identical to the previous version. There are 2 new test cases to cover some "skip message that wont fit" cases.